### DISCUSSION TITLE
problem with env variables

### DISCUSSION THREAD

**Discussion by i4lina** [dokku/dokku#5031] [Q&A]
Hi, i'm trying to deploy an app with doker on my newly created dokku server. I installed the postgres plugin, created a database and linked it to my app. For handling the database I'm using Prisma, which is an orm that can source a env variable containing the link to the database. On my pc I use a .env file containing this url, but as I'm trying to deploy I removed the .env file to let Prisma source the env variable for DATABASE_URL from the system. However, when I run the command to apply the migration, i get this error:
```
Error: Get config: Schema Parsing P1012

error: Environment variable not found: DATABASE_URL.
  -->  schema.prisma:7
   | 
 6 |   provider = "postgresql"
 7 |   url      = env("DATABASE_URL")
   | 

Validation Error Count: 1

error Command failed with exit code 1.
```
Is there a way to check if the right env variables are present when I try to deploy? I checked with dokku config:show <appname> and the DATABASE_URL variable is present.

**Comment by josegonzalez** [comment]
Can you show the output of the deploy? I'm assuming the migration happens in a hook. Also, what version of dokku are you running?

**Reply by i4lina** [reply]
When i try to push my repository to dokku I get this output: 
```
Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Delta compression using up to 8 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 332 bytes | 332.00 KiB/s, done.
Total 3 (delta 2), reused 0 (delta 0), pack-reused 0
-----> Set main to DOKKU_DEPLOY_BRANCH.
-----> Cleaning up...
-----> Building api from Dockerfile
remote: build context to Docker daemon  11.29MB
Step 1/17 : FROM node:16
 ---> 304de6a23023
Step 2/17 : WORKDIR /usr/src/app
 ---> Using cache
 ---> eaa597db1052
Step 3/17 : COPY package.json ./
 ---> Using cache
 ---> 2d0b1cd2e1bc
Step 4/17 : COPY yarn.lock ./
 ---> Using cache
 ---> 2263b443d033
Step 5/17 : RUN yarn
 ---> Using cache
 ---> b348916d1d74
Step 6/17 : COPY .env.production .env
 ---> Using cache
 ---> dc85e444210b
Step 7/17 : COPY . .
 ---> 4f5fff646e4e
Step 8/17 : RUN yarn prisma generate
 ---> Running in 146721ff97e7
yarn run v1.22.15
$ /usr/src/app/node_modules/.bin/prisma generate
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma

✔ Generated Prisma Client (3.8.0 | library) to ./node_modules/@prisma/client in 217ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
``
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()
``

warn Versions of prisma@3.9.1 and @prisma/client@3.8.0 don't match.
This might lead to unexpected behavior.
Please make sure they have the same version.
Done in 6.32s.
Removing intermediate container 146721ff97e7
 ---> 993d12ec9d4c
Step 9/17 : RUN yarn build
 ---> Running in 32da729e1248
yarn run v1.22.15
$ tsc
Done in 12.69s.
Removing intermediate container 32da729e1248
 ---> bf6337c67151
Step 10/17 : RUN yarn prisma migrate deploy
 ---> Running in 2e6a1ce94e61
yarn run v1.22.15
$ /usr/src/app/node_modules/.bin/prisma migrate deploy
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Error: Get config: Schema Parsing P1012

error: Environment variable not found: DATABASE_URL.
  -->  schema.prisma:7
   | 
 6 |   provider = "postgresql"
 7 |   url      = env("DATABASE_URL")
   | 

Validation Error Count: 1

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
remote: The command '/bin/sh -c yarn prisma migrate deploy' returned a non-zero code: 1
remote: 2022/02/14 11:27:30 exit status 1
To dokku.postspicker.com:api
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to 'dokku.redacted.com:api'
```
I'm running `dokku version 0.26.8` on `Ubuntu 20.04.3 LTS`

I temporarily solved the problem creating a Procfile and adding the command that failed during the deploy as a release phase command:
`release: yarn prisma migrate deploy`, and it works. This means that when the release script are executed the env variable is present. Do you think is possible to access the env variable in the docker build phase? So that I don't need to use a Procfile? Thanks for your response

**Reply by rohanrajpal** [reply]
+1 on this, going through the exactly same issue and trying to figure out how to access the postgres instance during build phase.

there is an `expose` command in postgres but i'm not sure what the dsn of postgres is after exposing it

**Reply by josegonzalez** [reply]
You _really_ shouldn't run migrations during a build process, and instead use a deployment task (either in app.json or in release). A good reason why is that CI should be able to build your images, and your CI process may not always be able to talk to your database (or you may wish to execute the build separately from an actual deployment where the migration would be applied).

That said, you can use the `--build-arg` docker-options flag to set custom env vars for your `Dockerfile` build. I updated the env var docs [here](https://github.com/dokku/dokku/pull/5230) to point to those docs in our [dockerfile deployment](https://dokku.com/docs/deployment/builders/dockerfiles/#build-time-configuration-variables) docs. Hope that helps.
